statemachine/statemachine.go

```diff
@@ -0,0 +1,316 @@
+	// between states.
+	Release MessageCode = "release"
+)
+
+// Transitions represent a state machine transition from one state to another
+// given an event message.
+type Transition struct {
+	Event MessageCode
+	From  StateCode
+	To    StateCode
+}
+
+func (t Transition) String() string {
+	return fmt.Sprintf("%v---%v--->%v", t.From, t.Event, t.To)
+}
+
+var (
+	// Rules is the state transition table.
+	Rules = []Transition{
+		// Runnable can transition to anything
+		{Event: Checkpoint, From: Runnable, To: Runnable},
+		{Event: Release, From: Runnable, To: Runnable},
+		{Event: Sleep, From: Runnable, To: Sleeping},
+		{Event: Complete, From: Runnable, To: Completed},
+		{Event: Kill, From: Runnable, To: Killed},
+		{Event: Error, From: Runnable, To: Fault},
+		{Event: Pause, From: Runnable, To: Paused},
+
+		// Sleeping can return to Runnable or be Killed/Paused
+		{Event: Checkpoint, From: Sleeping, To: Sleeping},
+		{Event: Sleep, From: Sleeping, To: Sleeping},
+		{Event: Run, From: Sleeping, To: Runnable},
+		{Event: Kill, From: Sleeping, To: Killed},
+		{Event: Pause, From: Sleeping, To: Paused},
+
+		// The error state transitions to either sleeping, failed, or released (to
+		// allow custom error handlers to workaround localitly related errors).
+		{Event: Sleep, From: Fault, To: Sleeping},
+		{Event: Error, From: Fault, To: Failed},
+
+		// Paused can return to Runnable, be put to Sleep, or Killed
+		{Event: Checkpoint, From: Paused, To: Paused},
+		{Event: Release, From: Paused, To: Paused},
+		{Event: Run, From: Paused, To: Runnable},
+		{Event: Sleep, From: Paused, To: Sleeping},
+		{Event: Kill, From: Paused, To: Killed},
+
+		// Completed, Failed, and Killed are terminal states that cannot transition
+		// to anything.
+	}
+)
+
+// StatefulHandler is the function signature that the state machine is able to
+// run. Instead of metafora.Handler's Stop method, StatefulHandlers receive
+// Messages via the commands chan and return their exit status via a Message.
+//
+// Normally StatefulHandlers simply return a Message as soon as it's received
+// on the commands chan. However, it's also acceptable for a handler to return
+// a different Message. For example if it encounters an error during shutdown,
+// it may choose to return that error as an Error Message as opposed to the
+// original command.
+type StatefulHandler func(taskID string, commands <-chan Message) Message
+
+type stateMachine struct {
+	taskID     string
+	h          StatefulHandler
+	ss         StateStore
+	cl         CommandListener
+	cmds       chan Message
+	errHandler ErrHandler
+}
+
+// New handler that creates a state machine and exposes state transitions to
+// the given handler by calling its Transition method. It should be created in
+// the HandlerFunc you use with metafora's Consumer.
+//
+// If ErrHandler is nil DefaultErrHandler will be used.
+func New(tid string, h StatefulHandler, ss StateStore, cl CommandListener, e ErrHandler) metafora.Handler {
```

Comment:
These things are kind of a pain to create, and it's not that obvious how to hook it up to the Consumer. Would love other ideas (although the implementation I've done internally isn't very much code... just expects people to set up a lot of bits and pieces).